Includes cleanup

pogof opened the pull request:
I have taken up the task of cleaning up the includes so everything is using only necessary headers.

As per suggestions from the Discord server Clang Build Analyzer (https://github.com/aras-p/ClangBuildAnalyzer) and Include What You Use (https://github.com/include-what-you-use/include-what-you-use/) have been run on the `beta` branch. The result of IWYU can be found here: https://github.com/pogof/itgmania-dev/blob/global-h-removal/IWYU-analysis.txt


There is a script included with IWYU that should try and fix the problems, however the analysis would like to do some absolute path includes which is not correct. So for now I will be doing it manually.

Some of the current include chains run very deep and have to be dealt with together so the code actually compiles. The analysis might need to be run multiple times after major changes.

I am only compiling for Linux locally. I already know there is an issue on Mac. Will deal with that bit later.

LUA has been includes as follows (or any combination of the 3 includes)
```
#ifndef _MSC_VER
extern "C"
{
#endif
#include "../extern/lua-5.1/src/lua.h"
#include "../extern/lua-5.1/src/lualib.h"
#include "../extern/lua-5.1/src/lauxlib.h"
#ifndef _MSC_VER
}
#endif
```
Ref: 
https://github.com/pogof/itgmania-dev/blob/81356b4bd10a68ba1d1694da973f8cc7a09864c6/src/LuaManager.h#L11C1-L20C7

sukibaby commented on src/LifeMeterBattery.h:
```diff
@@ -2,7 +2,22 @@
 #include "LifeMeter.h"
-#include "Sprite.h"
+
+#ifndef _MSC_VER
+extern "C"
+{
+#endif
+#include "../extern/lua-5.1/src/lua.h"
+#ifndef _MSC_VER
+}
+#endif
```
I think we should continue to use `LuaManager.h` for now; breaking out these chunks of `ifndef`s may be a bit much for the first pass here. 

(I already said this in Discord, but commenting here for posterity/anyone who didn't catch our conversation in the coding channel.)

pogof replied:
The block above has been replaced everywhere for `LuaManager.h` where it originated from. All 3 header files will be included, but better then to have it copied in multiple places.

Fixed
[17c06e7](https://github.com/itgmania/itgmania/pull/830/commits/17c06e79781ef3bd66796d23ecc918a0c4d17cbc)

sukibaby commented on src/global.h:
```diff
@@ -110,11 +110,11 @@ void ShowWarningOrTrace( const char *file, int line, const char *message, bool b
+//#include "RageException.h"
 
 // Call a function every `n` frames.
 // Each call site will get its own counter.
-#include <utility>
+//#include <utility>
 template <typename Func, typename... Args>
 void CallEveryNFrames(int n, Func&& f, Args&&... args) {
 	static int counter = 0;
```
Admittedly I probably should have moved this out of `global.h`, but forgot about it as soon as this made it in lol. Removing `<utility>` here, but requiring it to be included to make use of `CallEveryNFrames` seems awkward, so we can probably just move this function into its own header and then pull it in via that. 

I can do this though if you undo the change on line 117.

pogof replied:
Reverted.
[bfca800](https://github.com/itgmania/itgmania/pull/830/commits/bfca800868dafe07e848aabee26c9f3381757967)

sukibaby commented on src/ScreenOptionsManageCourses.cpp:
```diff
@@ -1,27 +1,41 @@
+
 #include "ScreenPrompt.h"
-#include "ScreenMiniMenu.h"
+
```
I wonder why these includes were here? The course manager screen should probably be tested with these removed to make sure these includes aren't needed.

pogof replied:
Was able to select marathon, all text seemingly as it should be. As far as "pick a marathon, play a marathon" it works.

Will leave open till I know what screen it actually is.

pogof replied:
CrashCringle says the screen itself is currently unused.

The header file is not referenced anywhere directly so I would tend to agree.

sukibaby commented on src/SongManager.h:
```diff
@@ -5,23 +5,34 @@ class LoadingWindow;
+#include <unordered_map>
 #include <vector>
+#include <algorithm>                                               // for copy
+#include <map>                                                     // for map
+#include <set>                                                     // for set
```
these comments are kinda unneeded

pogof replied:
[1391c5e](https://github.com/itgmania/itgmania/pull/830/commits/1391c5e7dc5271e7f9e4ad1287d6b1adc3afc015)

sukibaby commented on src/ScreenGameplay.h:
```diff
@@ -17,10 +16,28 @@
 
-class LyricsLoader;
 class ActiveAttackList;
 class CombinedLifeMeter;
```
it should probably also be tested with a song that has a `.lrc` file to be sure this isn't a regression

(i have a few, if needed)

pogof replied:
In the current state lyrics are working.

![image](https://github.com/user-attachments/assets/51b0152a-a684-4edb-b55f-9327e1f78c5b)
![image](https://github.com/user-attachments/assets/4761b35f-04fd-40ea-9637-819153773cef)

sukibaby commented on src/ScreenEdit.cpp:
```diff
@@ -37,12 +35,35 @@
+#include "Difficulty.h"                  // for DifficultyToString, Difficulty
+#include "IniFile.h"                     // for IniFile
+#include "InputFilter.h"                 // for INPUTFILTER, InputFilter
+#include "MessageManager.h"              // for BroadcastOnChangePtr, Broadc...
+#include "Preference.h"                  // for Preference, Preference1D
+#include "RadarValues.h"                 // for RadarValues
+#include "RageUtil.h"                    // for ssprintf, StringToFloat, Qua...
```
these are cut off? here and below (64-66)

pogof replied:
[1391c5e](https://github.com/itgmania/itgmania/pull/830/commits/1391c5e7dc5271e7f9e4ad1287d6b1adc3afc015)

sukibaby commented on src/MusicWheel.cpp:
```diff
@@ -14,19 +14,35 @@
 #include "MessageManager.h"
 #include "LocalizedString.h"
+#include "Command.h"           // for ParseCommands
+#include "Difficulty.h"        // for Difficulty, Difficulty_Invalid, Diffic...
```
another truncated comment

pogof replied:
[1391c5e](https://github.com/itgmania/itgmania/pull/830/commits/1391c5e7dc5271e7f9e4ad1287d6b1adc3afc015)

sukibaby commented on src/GameCommand.cpp:
```diff
@@ -23,6 +23,27 @@
+#include "Preference.h"        // for Preference, IPreference, Preference1D
+#include "RageUtil_AutoPtr.h"  // for REGISTER_CLASS_TRAITS
+#include "SongOptions.h"       // for SongOptions
+#include "SongUtil.h"          // for GetStepsByDescription, GetStepsByDiffi...
+#include "StdString.h"         // for operator+, CompareNoCase, CStdStr<>::M...
```
few more here

pogof replied:
[1391c5e](https://github.com/itgmania/itgmania/pull/830/commits/1391c5e7dc5271e7f9e4ad1287d6b1adc3afc015)

pogof committed:
removed #include "RageException.h" from global. Added anywhere where needed

pogof committed:
removed #include "RageException.h" from global. Added anywhere where needed

pogof committed:
Include cleanup in Stepmania.h/.cpp

pogof committed:
Include cleanup in Stepmania.h/.cpp

pogof committed:
include cleanup in (something)File.h/cpp

pogof committed:
include cleanup in (something)File.h/cpp

pogof committed:
include cleanup unix crashHandler

fix mac

pogof committed:
include cleanup unix crashHandler

fix mac

pogof committed:
include cleanup: Untangling starting from steps.h

pogof committed:
include cleanup: Untangling starting from steps.h

sukibaby commented on src/Main.cpp:
```diff
@@ -1,4 +1,3 @@
-#include "global.h"
```
Can't build on Windows without `global.h` included here. Visual Studio says there's no references to `global.h` here, but still.

![Screenshot 2025-05-20 171104](https://github.com/user-attachments/assets/6fb1ed0e-8f71-4378-96e4-405b469b885d)

Otherwise everything seems to work fine,  all I did was launch the game and play one song on keyboard, but everything worked as expected.

pogof replied:
Include has been returned. This will have to be resolved when removing global.h

sukibaby commented on src/Player.cpp:
```diff
@@ -36,14 +30,40 @@
 #include <cstddef>
 #include <vector>
-
+#include <limits.h>
+#include <stdlib.h>
```
Visual Studio says there are 0 references to these (also algorithm and memory) and I can seemingly build and run the game without them. Mainly curious if gcc or clang wants these here.

pogof replied:
compiles for me with clang. clang also doesn't seem to care much :shrug:

bwaggone replied:
On inspection I see std::max in this file, which is what algorithm is for. I presume the other includes are for similar reasons. Even if you can build without the include, IWYU is as much about removing bad includes as it is explicitly including what was implicit before.

pogof replied:
You are right, I misunderstood the first comment.

pogof commented:
With the power of Copilot I have figured out how to Git. I have pinged @sukibaby on Discord about this, but figured that I might as well just do it. 

I have went back and somewhat loosely redid what I already presented, but now the commits make logical sense, each should be somewhat reversible on its own and the PR should compile without errors on all platforms. All previous reviews have been addressed.

The nightly Action always fails but that was already happening before I started.

New work has been done only really in these two commits
f514bcdc7844bf565512a5454dc678ac12af6c66
979a83bdc458da3789eb25bdfa921bfbad0f65c8

sukibaby commented on src/global.h:
```diff
@@ -22,7 +22,6 @@
 #endif
 
 /* Make sure everyone has min and max: */
```
Unneeded comment

sukibaby commented on src/Steps.h:
```diff
@@ -4,20 +4,22 @@
+class Song;
+struct ColumnCue;
 class NoteData;
-struct lua_State;
```
I dunno, should we touch structs in this PR too? i'm leaning towards not touching it.  opinions from others?

pogof replied:
I think this is part of the forward declaration discussed above. Most likely will revert

aeubanks commented on src/EditModePlayerManager.cpp:
```diff
@@ -7,7 +7,20 @@
+
+#include <utility>
+
+class NoteData;
```
there are tradeoffs with forward declarations and I don't think itgmania uses them, I think for simplicity it would be nice to just include everything. is there an iwyu configuration to disable forward declarations?

pogof replied:
> Add --no_fwd_decls option to avoid replacing includes with forward-declarations

Seems like there is. I can run it with the option, but I would also like to hear more opinions.

bwaggone replied:
There are a couple places that use forward declarations already, but I agree with @aeubanks that we probably should just include everything explicitly.

The google style guide recommends avoiding them.
https://google.github.io/styleguide/cppguide.html#Forward_Declarations

teejusb replied:
No forward declaration is 100% preferred.

pogof replied:
Noted will address

pogof committed:
include cleanup: Add missing RageUtil headers in multiple files

pogof committed:
include cleanup: Add missing RageUtil headers in multiple files

pogof committed:
Stuff that broke when this branch was rebased to 1.3.0 beta branch

pogof committed:
Stuff that broke when this branch was rebased to 1.3.0 beta branch

pogof committed:
Include cleanup: another batch + windows fix

pogof committed:
Include cleanup: another batch + windows fix

pogof committed:
Include cleanup: Another 31 files cleaned

pogof committed:
Include cleanup: Another 31 files cleaned

pogof committed:
Include Cleanup: another 32 files

pogof committed:
Include Cleanup: another 32 files

pogof committed:
Include cleanup: another 27 files cleaned

pogof committed:
Include cleanup: another 27 files cleaned

pogof committed:
Include cleanup: yet another 31 done

pogof committed:
Include cleanup: yet another 31 done

pogof committed:
Include cleanup: 35 files cleaned

pogof committed:
Include cleanup: 35 files cleaned

pogof committed:
Include cleanup: 30 files more

pogof committed:
Include cleanup: 30 files more

pogof committed:
Include cleanup: 33 files more

pogof committed:
Include cleanup: 33 files more

pogof committed:
Include cleanup: 39 more files

pogof committed:
Include cleanup: 39 more files

pogof committed:
Inclue cleanup: 41 more files

pogof committed:
Inclue cleanup: 41 more files

pogof committed:
Include cleanup: 24 more files

pogof committed:
Include cleanup: 24 more files

pogof committed:
Include cleanup: yet another 35 files

pogof committed:
Include cleanup: yet another 35 files

pogof committed:
Include cleanup: 38 files more cleaned

pogof committed:
Include cleanup: 38 files more cleaned

pogof committed:
Include cleanup: 33 more files done

pogof committed:
Include cleanup: 33 more files done

pogof committed:
Include cleanup: yet another 24 files cleaned

pogof committed:
Include cleanup: yet another 24 files cleaned

pogof committed:
Include cleanup: 37 more files

pogof committed:
Include cleanup: 37 more files

pogof committed:
Include cleanup: 44 more files cleaned

pogof committed:
Include cleanup: 44 more files cleaned

pogof committed:
Include cleanup: Last 38 files for the initial sweep

pogof committed:
Include cleanup: Last 38 files for the initial sweep

pogof committed:
Include cleanup: clean-up sweep-up

pogof committed:
Include cleanup: clean-up sweep-up

pogof commented:
I have finally found some time and tried to do at least the initial sweep. I initially wanted to do this in chunks, however after trying to do it in the first few commits somewhat thematically, it became obvious that this just wasn't happening. After running IWYU and just going in order of compilation it actually wasn't as difficult to do as I initially thought, though we lost any sense of theme for the commits. Since this has been up for almost a year now I had to rebase this multiple times, each time messing it up a bit and having to fix it.

## What has been done any why:
I have inserted IWYU into cmake and let it generate a report. I have disabled comments and made sure forward declarations were not used.
`CC=clang CXX=clang++ cmake -DCMAKE_CXX_INCLUDE_WHAT_YOU_USE="include-what-you-use;-Xiwyu;--no_fwd_decls;-Xiwyu;--no_comments" ..`

I have manually went and added or removed includes in files that required it. I have, where it wasn't obvious, tried to make sure that any removals/additions make sense. I have checked and made sure that the game will compile without issues via the repo actions. In the current state it compiles on all platforms without any errors.

### Lua inclusion
Initially I wanted to include the extern Lua dependencies as stated in the OP, however after realizing at how many places they are required, I have opted for a slimmed down version of including `LuaManager.h` where they were necessary.

### Misc stuff
IWYU sometimes wanted me to include stuff like `alloca.h`, `strings.h`, which will not compile on Windows, however it compiled fine even without explicitly including them. I would assume these probably should be included but only if POSIX is available.

Some files have not been touched or just very sporadically as it wasn't in best interest messing with them. Primarily crypt and input io stuff.

### Forward declarations, structs, etc.
 Nothing has been touched, added or removed. IWYU is telling me there is a lot that probably could be removed. However I am not sure if it should still be in the scope of this PR. 

I have attached the log that has been generated while compiling the game at the latest commit of this PR.

[stderr.log](https://github.com/user-attachments/files/26491580/stderr.log)

sukibaby commented on src/arch/InputHandler/InputHandler_DirectInput.cpp:
```diff
@@ -12,7 +26,6 @@
 #include "archutils/Win32/ErrorStrings.h"
 #include "archutils/Win32/GraphicsWindow.h"
 #include "archutils/Win32/RegistryAccess.h"
-#include "global.h"
```
This is odd to see... was a reason given for this?

pogof replied:
Aha, I finally figured out whats going on. As I already stated (and most likely will few more times) this is yet another victim of the messed up rebase. I think I have moved the includes around and this just got lost? It is windows specific file, I have not compiled it myself on Win, ~~that shouldn't really have been touched at all.~~
Had to be touched, as changes elsewhere made it not work, but `#include global.h` was restored

sukibaby commented on src/arch/InputHandler/InputHandler_Linux_PIUIO.h:
```diff
@@ -1,10 +1,7 @@
-#include <vector>
-
 #include "InputHandler.h"
-#include "RageInputDevice.h"
```
Curious why this comes up in so many files if it's not needed. Should probably test the Linux input stuff more heavily to ensure it all behaves the same.

pogof replied:
It should just be these 4 that lost it.
* `InputHandler_PumpHID.h`
*  `InputHandler_Linux_Joystick.h`
* `InputHandler_Linux_PIUIO.h`
* `InputHandler_SextetStream.h`

The common denominator is `InputDeviceInfo` which is defined in `RageInputDevice.h` however it is only ever used with the function `virtual void GetDevicesAndDescriptions()` that comes from `InputHandler.h` and that already includes `RageInputDevice.h`, making it in my opinion redundant.

sukibaby commented on src/arch/Lights/LightsDriver_gpb.cpp:
```diff
@@ -1,9 +1,10 @@
 #include "LightsDriver_gpb.h"
 
-#include "Game.h"
-#include "GameState.h"
-#include "RageLog.h"
-#include "global.h"
```
curious why the includes are changing so much on this one,  din committed this not too long ago so just curious why the includes are changing so much? was it necessary or recommended?

pogof replied:
IWYU if compiled with the comments on will tell you what the includes are used for. For this file it gives the following:
```
#include <string.h>           // for memcmp, memcpy
#include "GameInput.h"        // for GameController, GAME_BUTTON_DOWN, GAME_...
#include "LightsManager.h"    // for LightsState
#include "arch/RageDriver.h"  // for RegisterRageDriver
```
which as far as this file goes is literally everything used. Would there be any reason to include `global.h` and all of `GameState.h`?

sukibaby commented on src/arch/LowLevelWindow/LowLevelWindow_X11.cpp:
```diff
@@ -20,7 +35,7 @@ using namespace X11Helper;
 #include <string>
 #define GLX_GLXEXT_PROTOTYPES
-#include <GL/glx.h>  // All sorts of stuff...
+#include <GL/glx.h>  // All sorts of stuff... //Unused?
```
would prefer to find out if this is needed or not than add a comment indicating uncertainty

pogof replied:
I think it is made redundant by the inclusion of `#include <GL/glew.h>` and `#include <GL/glxew.h>`, but I'm leaving it in and putting a pin on this to double check.

pogof replied:
Removed, redundant by `GL/glxew.h`.

sukibaby commented on src/archutils/Unix/X11Helper.cpp:
```diff
@@ -1,13 +1,14 @@
 
 #include "Preference.h"
-#include "PrefsManager.h"
+#include "PrefsManager.h"  // XXX: only used for m_bShowMouseCursor -aj
```
this isn't actually aj's comment is it ??

pogof replied:
This actually was removed here: 1c220739656ef211c5a9a126a5237f5a9ac56779

This got back when I fucked up rebase. Will remove.

sukibaby commented on src/GameManager.cpp:
```diff
@@ -7,19 +7,19 @@
 #include "Game.h"
 #include "GameConstantsAndTypes.h"
-#include "GameInput.h"  // for GameButton constants
-#include "GameLoop.h"   // for ChangeGame
```
from a quick check, these comments seem to not be needed for context so it seems ok to remove them

sukibaby commented on src/global.h:
```diff
@@ -17,8 +17,6 @@
 #include "archutils/Win32/arch_setup.h"
 #elif defined(PBBUILD) || defined(MACOSX)
 #include "archutils/Darwin/arch_setup.h"
-#elif defined(UNIX)
-#include "archutils/Unix/arch_setup.h"
 #endif
```
This seems odd, even though on my Linux system it still chose Pulse / opengl / etc correctly, but if so  I don't immediately see why the Unix arch_setup.h wouldn't be needed...

pogof replied:
Not strictly needed for any remotely modern Linux system, however could cause problems on older systems or for BSD systems. Reverted.

sukibaby commented on src/Player.cpp:
```diff
@@ -1,7 +1,10 @@
+
 #include <algorithm>
-#include <climits>
+#include <bitset>
```
This is odd to see. I'm curious if it's misreading a function in here as being from `<bitset>` or if it's genuine.

pogof replied:
This yet again most likely disappeared in the messed up rebase. IWYU says it should be there.  Reverted.

sukibaby commented on src/ScoreKeeperNormal.cpp:
```diff
@@ -7,28 +7,31 @@
 #include <string>
 #include <vector>
 
-#include "Course.h"
```
I wonder why it's being removed if it was the first one? Just not sure if that was intentional or not, so this might be worth a double check to make sure there's not some ambiguous polymorphism or anything of the sort.

pogof replied:
Just lightly skimming through the file only `GAMESTATE->IsCourseMode()` is used and that comes from `GameState.h`

I am not 100% sure if your comment was supposed to be relevant to this or something else?

sukibaby commented on src/ScreenOptionsExportPackage.cpp:
```diff
@@ -17,7 +18,7 @@
 #include "SpecialFiles.h"
 #include "StdString.h"
-#include "archutils/SpecialDirs.h"
+#include "archutils/Unix/SpecialDirs.h"
```
this doesn't seem right - though tbh it's probably worth deprecating `archutils/Unix/SpecialDirs.h"` after looking at it

pogof replied:
I agree. Reverted.

sukibaby commented on src/ScreenPrompt.h:
```diff
@@ -7,7 +7,6 @@
 
 #include "AutoActor.h"
 #include "BitmapText.h"
-#include "InputQueue.h"
```
This seems to be getting removed from a lot of places, just curious why. Not familiar with this header specifically.

pogof replied:
As far as I was able to understand this, `InputQueue.h` is involved in detection of multiple menu button presses/holds for entering codes/shortcuts/cheat codes. It is used within `CodeDetector.cpp`, the codes are defined in theme metrics. It looks like most headers either don't take advantage of this or it was used to pass through`#include "InputEventPlus.h"` which is more appropriate.

sukibaby commented on src/SongManager.cpp:
```diff
@@ -22,16 +18,14 @@
 #include "MemoryCardManager.h"
+#include "MessageManager.h"
 #include "MsdFile.h"
-#include "NotesLoaderDWI.h"
```
That's unexpected I think?

pogof replied:
After skimming through the file, DWI is just not referenced anywhere but SM and SSC very much are. As far as the scope of this PR goes `NotesLoader.DWI.h` is validly unused. Probably should be looked into, but it is out of the scope of this PR.

sukibaby commented on src/TrailUtil.cpp:
```diff
@@ -1,13 +1,10 @@
+#include <vector>
 
 #include "Course.h"
-#include "Difficulty.h"
```
huh.... i wonder if Trail.h/.cpp is used even

sukibaby reviewed (commented):
Okay, this took forever to get through, lol. I've been trying to finish getting thru this review for a while now and finally persevered. 

There is not a lot to say in regards to changed includes, so I mostly commented on a few that stuck out to me as strange or interesting...

I did pull the branch, build it, and did initial informal testing, where nothing was obviously broken. I haven't done heavy testing of course, but on a surface level it looks like things are working fine.

pogof commented:
Thanks Jasmine for the code review. All concerns have been addressed and resolved by either reverting or justifying the removal/addition. 

I do not own a mac and I am not familiar enough with the windows workflow to make any big changes + analysis on files relevant to just those (+ this PR is really leverages IWYU and clang).

Unless there are any other concerns I would deem this as ready for merging. I will try to upkeep this to the beta branch.

I also assume this will get squashed at some point. I have not done this yet, but it might get increasingly difficult to upkeep in the current state.

teejusb commented:
Merge conflicts but generally seems good to me.

pogof commented:
I was just yesterday evening looking if this needs to be rebased, doing so now.

pogof committed:
Fixing errors after rebase

aeubanks commented on src/arch/InputHandler/InputHandler_MonkeyKeyboard.h:
```diff
@@ -1,8 +1,6 @@
 #ifndef INPUT_HANDLER_MONKEY_SCRIPT
 #define INPUT_HANDLER_MONKEY_SCRIPT
 
-#include <vector>
```
spot checking, this seems wrong, we do reference std::vector in this header

pogof committed:
Include cleanup: updates after rebasing

pogof commented:
@teejusb Rebased and new stuff cleaned too.
